### PR TITLE
SONARJAVA-1604 FN on S3066 - Public final fields should be checked for their mutability

### PR DESCRIPTION
**Pull request by fergarrui**
* Now `public` `final `**mutable** fields are raising an issue.
* Mutable fields considered: `java.util.Collections`, `java.util.Date` and arrays.
* Collections extending [com.google.common.collect.ImmutableCollection](https://google.github.io/guava/releases/snapshot/api/docs/com/google/common/collect/ImmutableCollection.html) are not considered mutable in the rule.

### CODE REVIEW

**Comment by m-g-sonar** [comment]
Hello @fergarrui,

Thanks for your contribution. This JIRA ticket is planned for version 4.4 of the Java Analyzer. So I'll review in details after release of 4.3!

Cheers,
Michael

File: java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java [hunk 1]
```diff
@@ -65,6 +63,24 @@ public void visitNode(Tree tree) {
...
+  private static boolean isMutableFinalMember(VariableTree variableTree) {
+    return ModifiersUtils.hasModifier(variableTree.modifiers(), Modifier.FINAL) && isMutableMember(variableTree);
+  }
+  
+  private static boolean isMutableMember(VariableTree variableTree) {
+    return variableTree.type().is(Tree.Kind.ARRAY_TYPE) || isDateOrCollection(variableTree);
+  }
+  
+  private static boolean isDateOrCollection(VariableTree variableTree) {
+    return variableTree.symbol().type().is("java.util.Date") ||
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java]: extract the type as `variableTree.symbol().type()` in order to not have to call it 3 times in worst cases.

**Comment by fergarrui** [comment]
Changes done!

File: java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java [hunk 2]
```diff
@@ -65,6 +64,25 @@ public void visitNode(Tree tree) {
       }
     }
   }
+  
+  private static boolean isNotStaticOrFinal(VariableTree variableTree, ModifiersTree modifiersTree) {
```
Comment (sonartech-externalpr) [inline java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java]: MAJOR Remove this unused method parameter "variableTree". [rule](https://sonarqube.com/coding_rules#rule_key=squid%3AS1172)

**Comment by sonartech-externalpr** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java (hunk 2 above)
Comment (fergarrui) [inline java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java]: Not detecting my last push?

commit 4e10991a424e0c1f20fd2d574ad5af51fd80ef36
Author: Fernando Garcia <fergarrui@gmail.com>
Date:   Tue Dec 6 20:33:01 2016 +0000

    SONARJAVA-1604 - fix sonar issue

commit 2f4b61ce0b82358a4c5d79ae35302355d71b5394
Author: Fernando Garcia <fergarrui@gmail.com>
Date:   Tue Dec 6 19:03:13 2016 +0000

    SONARJAVA-1604 - Requested changes

Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/EnumMutableFieldCheck.java]: Indeed, your commit fixing the sonar issue is apparently not part of the PR. No idea why,

**Comment by fergarrui** [comment]
I did a dummy push and it added the push from yesterday  and this one.
Then I reverted it (sorry to mess up). That's so weird, but now it's fine.

**Comment by m-g-sonar** [comment]
No worries, everything LGTM now.
Thanks for the udpate! :)
